trace exec: fix sys_exit_execve from thread

alban opened the pull request:
sys_enter_execve adds an entry in the "execs" map and sys_exit_execve removes it. However, sys_enter_execve and sys_exit_execve can be called from different thread id in case of a multithreaded application. This caused the map to not get cleaned up properly.

This patch fixes the tracking between sys_enter_execve and sys_exit_execve.

Fixes https://github.com/inspektor-gadget/inspektor-gadget/issues/2424

## How to use

No changes

## Testing done

- Workload from in #2424

```
$ go run -exec 'sudo -E' ./cmd/ig/... trace exec
RUNTIME.CONTAINERNAME            PID        PPID       COMM              RET ARGS                                     
goexecve                         153115     153094     sh                0   /usr/bin/sh -c for i in $(seq 5); do /op…
goexecve                         153140     153115     seq               0   /usr/bin/seq 5                           
goexecve                         153141     153115     goexecve          0   /opt/goexecve                            
goexecve                         153141     153115     cat               0   /bin/cat /dev/null                       
goexecve                         153148     153115     goexecve          0   /opt/goexecve                            
goexecve                         153148     153115     cat               0   /bin/cat /dev/null
...                   
```

```
$ sudo bpftool map dump name pid_by_tgid
[]
$ sudo bpftool map dump name execs
[]
```

- Workload that runs 2 execve in parallel from 2 threads (using a fuse filesystem to suspend an execve in the middle as long as desired)

```
$ go run -exec 'sudo -E' ./cmd/ig/... trace exec
RUNTIME.CONTAINERNAME            PID        PPID       COMM              RET ARGS                                     
goexecve                         153473     153452     sh                0   /usr/bin/sh -c cd /opt && /opt/ex        
goexecve                         153497     153473     ex                0   /opt/ex                                  
goexecve                         153497     153473     ex                -2  ./script                                 
goexecve                         153497     153473     true              0   /bin/true
```

While the 2 execve are suspended by the fuse filesystem:
```
$ sudo bpftool map dump name pid_by_tgid
[{
        "key": 153497,
        "value": 306995
    }
]
$ sudo bpftool map dump name execs|grep key
        "key": 153498,
        "key": 153497,
```

After fuse I/O resumed:

```
$ sudo bpftool map dump name execs
[]
$ sudo bpftool map dump name pid_by_tgid
[]
```

mauriciovasquezbernal commented on pkg/gadgets/trace/exec/tracer/bpf/execsnoop.bpf.c:
```diff
@@ -181,10 +223,35 @@ int ig_execve_x(struct trace_event_raw_sys_exit *ctx)
+	if (!pid_map)
+		return 0;
+
+	if (ret == 0) {
+		// sys_enter_execve and sys_exit_execve might be called from different
+		// threads. We need to lookup the pid from the tgid.
+		execs_lookup_key = *pid_map;
+		event = bpf_map_lookup_elem(&execs, &execs_lookup_key);
+
+		// Remove the tgid->pid mapping. This was the last thread running a
+		// execve in parallel since there could be only one successful execve
+		// in the thread group.
+		bpf_map_delete_elem(&pid_by_tgid, &tgid);
+	} else {
+		execs_lookup_key = pid;
+		event = bpf_map_lookup_elem(&execs, &execs_lookup_key);
+		// Remove the tgid->pid mapping.
+		__atomic_add_fetch(pid_map, -pid, __ATOMIC_RELAXED);
+		if (*pid_map == 0) {
+			bpf_map_delete_elem(&pid_by_tgid, &tgid);
+		}
+	}
+
```
I'm wondering if something like the following is more readable:

```suggestion
	// sys_enter_execve and sys_exit_execve might be called from different
	// threads. We need to lookup the pid from the tgid.
	execs_lookup_key = (ret == 0) ? *pid_map : pid;

	// Remove the tgid->pid mapping if the value reaches 0 or the execve() call was successful
	__atomic_add_fetch(pid_map, -pid, __ATOMIC_RELAXED);
	if (*pid_map == 0|| ret == 0) {
		bpf_map_delete_elem(&pid_by_tgid, &tgid);
	}

	event = bpf_map_lookup_elem(&execs, &execs_lookup_key);
```

mauriciovasquezbernal reviewed (approved):
LGTM. I have been trying to think about a simpler solution but I haven't been able.

alban commented on pkg/gadgets/trace/exec/tracer/bpf/execsnoop.bpf.c:
```diff
@@ -38,6 +38,36 @@ struct {
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, pid_t); // tgid
+	__type(value, pid_t); // pid
```
I was thinking that overflow was fine because that would just do the computation modulo `2^32`: when adding and removing the same pid, we fall back on the previous value, even when performing all operations modulo `2^32`, and we only read the value once there is one pid remaining.

However, we do the test `if (*pid_map == 0)` and there is a risk that several pids cancel each other out when adding them modulo `2^32`. So I will try to use u64 to avoid this problem.

alban commented:
> I implemented a unit test on https://github.com/inspektor-gadget/inspektor-gadget/commit/f701b1bb649158ddcb45586d5db6838d2fc0979c, feel free to pick it if you want. (I wasn't sure about using Python, but I didn't find another way to implement it)

Thanks, I picked your patch and I added a second unit test with it.

alban commented:
Follow-up for the image-based gadget: #2475
